### PR TITLE
[WIP] Allow compilation to in memory stream, part 1: Fsc output

### PR DESCRIPTION
**Pull request by enricosada**
this pr is the first of a multuple set to allow in memory compilation and output

see fsharp/FSharp.Compiler.Service#266
user voice: https://fslang.uservoice.com/forums/245727-f-language/suggestions/6717137-allow-for-completely-in-memory-compilation

the refactoring plan is to start with the lower level, convert paths to stream factories and bubble up these arguments up to fsc main.
There we can build a better api for in memory compilation ( really nice for testing or compiler service i think)

current fsc main:
```fsharp
    main0(argv,bannerAlreadyPrinted,exiter, errorLoggerProvider, disposables)
    |> main1
    |> main2
    |> main2b
    |> main2c
    |> main3 
    |> main4
```

inside `main0` the command line argument are parsed and outfile/pdbfile path is set and passed to all other funcions ( `main1` to `main4` )

this pr add output/pdb/mdb to stream or file. 
The output of assembly directly to memorystream works, pdb/mdb/signed require disk io with temp file.

the api to specify where send out/pdb/mdb is

```fsharp
type EmitTo =
    | EmittedFile of string
    | EmittedStream of Stream
``` 

with error and deferred creation

```fsharp
type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
type Diagnostic = string
```

1. `Choice<EmitTo, Diagnostic>`:
  -  to use success/fail pattern instead of exception based control flow
  - we can refactor (later) the control flow as computation expression  
2. `Lazy`:
  -  deferred creation
  - once
  - we can ask ( `Lazy.IsValueCreated` ) if the factory is called, and easy cleanup the files on error
3. both `EmittedFile of string` and `EmittedStream of Stream` , not only `Stream`
  - performance, useless continue to copy stream (build+sign, pdb) if i need a real file
  - current implementation (pdb/sign) require file on disk, useless hide that complexity. if we later improve pdb/sign, this api is ok 
  - easy to consume api

As side note, roslyn do pretty much the same with [EmitStreamProvider](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/Compilation.EmitStreamProvider.cs,0a7a5f6ff635682d,references), so we can wrap later when a bridge is ready ( working on a partial bridge :smile: )

i need a feedback about naming, i think is weak (for example `EmittedFile` -> `EmitToFile`? dunno)

### CODE REVIEW

**Comment by dsyme** [comment]
@enricosada To clarify - this looks like an FCS feature, but you'd like the core implementation diff that enables the feature to be integrated into Microsoft/visualfsharp to keep maximal code alignment?  If so I am assuming that you are confident that the feature is non-intrusive and non-risky.  Also, just to check, do you think it will commute well with the CoreCLR work?

@KevinRansom If so I'm ok with the approach of commiting the core of FCS implementation feature work to Microsoft/visualfsharp first, then integrating through to FCS.  If you like that approach, then we should probably bring over the core implementation diff of a couple of other FCS features like in-memory cross-project references too.  That will continue to put us in a really good position to fully align in the future (e.g. having the ability to build compatible FSharp.Compiler.Service DLLs out of Microsoft/visualfsharp).

@enricosada If @KevinRansom doesn't feel able to pull in the next couple of days for stability reasons, you can submit first to FCS.  Make sure the diff is absolutely minimal.

Thanks 
Don

File: src/absil/ilwrite.fs [hunk 1]
```diff
@@ -222,7 +222,6 @@ type PdbData =
 //---------------------------------------------------------------------
 
 let WritePdbInfo fixupOverlappingSequencePoints showTimes f fpdb info = 
-    (try FileSystem.FileDelete fpdb with _ -> ())
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: What's this removal for?

File: src/absil/ilwrite.fs [hunk 2]
```diff
@@ -3875,7 +3874,28 @@ let writeDirectory os dict =
...
-let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer: ILStrongNameSigner option, fixupOverlappingSequencePoints, emitTailcalls, showTimes, dumpDebugInfo) modul noDebugData =
+type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
+and EmitTo =
+    | EmittedFile of string
+    | EmittedStream of Stream
+and Diagnostic = string
+
+
+// Either
+let (|Success|Failure|) = function
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: This looks needless, just use Choice1Of2 and Choice2Of2

File: src/absil/ilwrite.fs [hunk 3]
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
           let entrypointCodeChunk,next = chunk 0x06 next
           let globalpointerCodeChunk,next = chunk (if isItanium then 0x8 else 0x0) next
           
-          let debugDirectoryChunk,next = chunk (if pdbfile = None then 0x0 else sizeof_IMAGE_DEBUG_DIRECTORY) next
+          let debugDirectoryChunk,next = chunk (if pdbfile = false then 0x0 else sizeof_IMAGE_DEBUG_DIRECTORY) next
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: use ``not pdbfile``

File: src/absil/ilwrite.fs [hunk 4]
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
           // The debug data is given to us by the PDB writer and appears to 
           // typically be the type of the data plus the PDB file name.  We fill 
-          // this in after we've written the binary. We approximate the size according 
+          // this in after we've written the binary. We approximate the size to MAXPATH according 
           // to what PDB writers seem to require and leave extra space just in case... 
           let debugDataJustInCase = 40
+          let MAXPATH = 260
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Where's this constant coming from?

File: src/absil/ilwrite.fs [hunk 5]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
-                writeInt32 os2 idd.iddData.Length  // IMAGE_DEBUG_DIRECTORY.SizeOfData 
-                writeInt32 os2 debugDataChunk.addr  // IMAGE_DEBUG_DIRECTORY.AddressOfRawData 
-                writeInt32 os2 (textV2P debugDataChunk.addr)// IMAGE_DEBUG_DIRECTORY.PointerToRawData 
-
-                (* dprintf "idd.iddCharacteristics = %ld\n" idd.iddCharacteristics
-                dprintf "iddMajorVersion = %ld\n" idd.iddMajorVersion
-                dprintf "iddMinorVersion = %ld\n" idd.iddMinorVersion
-                dprintf "iddType = %ld\n" idd.iddType
-                dprintf "iddData = (%A) = %s\n" idd.iddData (System.Text.Encoding.UTF8.GetString idd.iddData) *)
+            // write the IMAGE_DEBUG_DIRECTORY
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Please minimize this diff by replacing the try-with with a bing-end block and keeping the indentation the same

File: src/absil/ilwrite.fs [hunk 6]
```diff
@@ -4025,17 +4037,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
           let debugDataJustInCase = 40
+          let MAXPATH = 260
           let debugDataChunk,next = 
               chunk (align 0x4 (match pdbfile with 
-                                | None -> 0x0 
-                                | Some f -> (24 
-                                            + System.Text.Encoding.Unicode.GetByteCount(f) // See bug 748444
+                                | false -> 0x0 
+                                | true -> (24 
+                                            + MAXPATH
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: To get into Microsoft/visualfsharp you'll need to have written the code so that every point we can see that it's 100% "safe" - i.e. there is absolutely zero risk of behaviour change.  So for this kind of thing you will need need to take the max between MAXPATH and GetByteCount, or leave the behaviour exactly the same as before when using a on-disk file as output.

File: src/absil/ilwrite.fs [hunk 7]
```diff
@@ -4116,7 +4129,6 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
               match p with 
               | None -> () 
               | Some pExpected -> 
-                  os.Flush();
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: What's this removal and why is it safe?

File: src/absil/ilwrite.fs [hunk 8]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
                  b0 reloc2; b1 reloc2; |]
           writePadding os "end of .reloc" (imageEndSectionPhysLoc - relocSectionPhysLoc - relocSectionSize)
 
-          os.Close()
-          
-          try 
-              FileSystemUtilites.setExecutablePermission outfile
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: The call to setExecutablePermission was surrounded by a try-with - has that been maintained in the code?  Again we need to see that there is zero chance of behaviour change for on-disk file emit. thanks

File: src/absil/ilwrite.fs [hunk 9]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
+        path |> FileSystemUtilites.setExecutablePermission 
+
+        pdb |> Option.iter (serializeToPdb (writePdb (pdbData, textV2P, debugDirectoryChunk, debugDataChunk) path))
+        mdb |> Option.iter (serializeToMdb (WriteMdbInfo pdbData path))
+        signer |> Option.iter (signTo path)
 
     mappings
 
 
+let private optionOfLazy (l: Lazy<'T>) =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Why is this needed?  Observing the is-this-evaluated state of lazy values is a bit of an anti-pattern and definitely best avoided

File: src/fsharp/fsc.fs [hunk 10]
```diff
@@ -1696,15 +1696,27 @@ module FileWriter =
                             with e -> 
                                 // Note:: don't use errorR here since we really want to fail and not produce a binary
                                 error(Error(FSComp.SR.fscKeyFileCouldNotBeOpened(s),rangeCmdArgs))
+
+                let toFile path = lazy(ILBinaryWriter.EmitTo.EmittedFile path |> Choice1Of2)
+                let toFileOpt = Option.map toFile
+                
+                let pdbfileOpt = if not(runningOnMono) then pdbfile |> toFileOpt else None
```
Comment (dsyme) [inline src/fsharp/fsc.fs]: No parens please after ``not``

File: src/absil/ilwrite.fsi [hunk 11]
```diff
@@ -15,17 +15,22 @@ type ILStrongNameSigner =
     static member OpenKeyPairFile: string -> ILStrongNameSigner
     static member OpenKeyContainer: string -> ILStrongNameSigner
 
+type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
```
Comment (dsyme) [inline src/absil/ilwrite.fsi]: I find using ``Lazy`` here is a bit awkward - in general it's just rarely a good type to expose in APIs.  I think I'd prefer just 

    [<RequireQualifiedAccess>]
    type EmitTo = 
        | File of string
        | Stream of (unit -> Stream)

    type options = 
        { .. pdbfile : EmitTo option ...  }    

Internally to ilwrite.fs you can convert to a register that amortizes the evaluation in the EmitTo case.  If tracking diagnostics is a problem then I'm OK with you using 

        | Stream of (unit -> Choice<Stream, exn>)

or

        | Stream of (unit -> Choice<Stream, Diagnostic>)

Could you see if that works?

File: src/absil/ilwrite.fs [hunk 12]
```diff
@@ -3875,7 +3874,28 @@ let writeDirectory os dict =
...
+let (|Success|Failure|) = function
+    | Choice1Of2 s -> Success s
+    | Choice2Of2 f -> Failure f
+
+let private succeed x = Choice1Of2 x
+let private fail x = Choice2Of2 x
+
+let private tryDeleteFile path = 
+    try
+        if FileSystem.SafeExists path
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: ``then`` at end of line please

    if ABC then
        DEF

File: src/absil/ilwrite.fs [hunk 13]
```diff
@@ -3916,15 +3936,7 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
-    let os = 
-        try  
-            new BinaryWriter(FileSystem.FileStreamCreateShim(outfile))
-        with e -> 
-            failwith ("Could not open file for writing (binary mode): " + outfile)    
-
-    let  pdbData,debugDirectoryChunk,debugDataChunk,textV2P,mappings =
-        try 
-      
+    let gen os (pdbfile: bool) =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Is it possible to avoid having a new inner function here?  It is only called in one place.   Why not put the prologue that computes "os" and "pdbfile" in the portion directly above this?  I'm looking to minimize the diff and keep thee current overall function structure the same

File: src/absil/ilwrite.fs [hunk 14]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
+            writeInt32 os2 idd.iddCharacteristics // IMAGE_DEBUG_DIRECTORY.Characteristics 
+            writeInt32 os2 timestamp
+            writeInt32AsUInt16 os2 idd.iddMajorVersion
+            writeInt32AsUInt16 os2 idd.iddMinorVersion
+            writeInt32 os2 idd.iddType
+            writeInt32 os2 idd.iddData.Length  // IMAGE_DEBUG_DIRECTORY.SizeOfData 
+            writeInt32 os2 debugDataChunk.addr  // IMAGE_DEBUG_DIRECTORY.AddressOfRawData 
+            writeInt32 os2 (textV2P debugDataChunk.addr)// IMAGE_DEBUG_DIRECTORY.PointerToRawData 
+
+            (* dprintf "idd.iddCharacteristics = %ld\n" idd.iddCharacteristics
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: You can delete this stuff

File: src/absil/ilwrite.fs [hunk 15]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
+            let tempFilePath = Path.ChangeExtension(FileSystem.GetTempFilePathShim(), extension)
+            tryDeleteFile tempFilePath
+            f tempFilePath
+            use tempFileStream = FileSystem.FileStreamReadShim(tempFilePath)
+            tempFileStream.CopyTo(stream)
+
+    let serializeToPdb = serializeToFile ".pdb"
+    let serializeToMdb = serializeToFile ".mdb"
+
+    let emitOrFail (e: EmitStreamProvider) =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Please pull this out to be a top-level function since it is independent

File: src/absil/ilwrite.fs [hunk 16]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
             failwith ("Warning: A call to StrongNameSignatureGeneration failed ("+e.Message+")")
-            (try s.Close() with _ -> ())
-            (try FileSystem.FileDelete outfile with _ -> ()) 
-            ()
-
-    reportTime showTimes "Signing Image"
-    //Finished writing and signing the binary and debug info...
+        reportTime showTimes "Signed Image"
+
+    let serializeToFile extension f emitter =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: This looks independent of context (no free variables) - please pull it out to be a top level function

File: src/absil/ilwrite.fs [hunk 17]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
-    match signer with 
-    | None -> ()
-    | Some s -> 
-        try 
-            s.SignFile outfile
-            s.Close() 
+        reportTime showTimes "Finalize PDB"
+
+    /// Sign the binary.  No further changes to binary allowed past this point!
+    let signTo filePath (s: ILStrongNameSigner) =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: This function looks independent of context (no free variables) - please pull it out to be a top-level function with a /// comment explaining what it does

File: src/absil/ilwrite.fs [hunk 18]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
-   
-
-    reportTime showTimes "Writing Image"
-     
-    if dumpDebugInfo then 
-        DumpDebugInfo outfile pdbData
+
+    let writeDumpDebugInfo pdbData dumpTo =
+        reportTime showTimes "Dump Debug Info"
+        let dump stream = DumpDebugInfo stream pdbData
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Please avoid little nested helper functions like this - inlining the call to DumpDebugInfo in two places is clearer to read than a nested inner function just called ``dump``.

File: src/absil/ilwrite.fs [hunk 19]
```diff
@@ -4462,105 +4474,146 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
-             with _ -> ()) 
-            reraise()
-   
-
-    reportTime showTimes "Writing Image"
-     
-    if dumpDebugInfo then 
-        DumpDebugInfo outfile pdbData
+
+    let writeDumpDebugInfo pdbData dumpTo =
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: Perhaps the content of this should just be combined with DumpDebugInfo

**Comment by dsyme** [comment]
Added a bunch of comments - thanks!  It's great to have this underway - FCS can really benefit from this feature.

File: src/absil/ilwrite.fs (hunk 1 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: removed because useless, the file is now removed in `serializeToFile` (the same for the .mdb)

File: src/absil/ilwrite.fs (hunk 2 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: About that, i think `Success`/`Failure` make the code easier to read instead of `Choice1Of2` / `Choice2Of2` .

i can declare a type

```fsharp
type Result<'S, 'F> = Success of 'S | Failure of 'F
```

but Choice is the same, with a little help of active pattern. But maybe the custom DU is better?

I really like to refactor to change the code from exception workflow -> Success/Failure -> computation expression because is times easier to read and is easier to use

File: src/absil/ilwrite.fs (hunk 4 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: this is windows [MAX_PATH](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396#maxpath), the maximum length of a file path

File: src/absil/ilwrite.fs (hunk 6 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: previous: size variable, but the upper bound is MAX_PATH
now: directly MAX_PATH 

is pretty much the same, but i dont need to know the pdb file name (because help a lot the refactoring)

File: src/absil/ilwrite.fs (hunk 7 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: `os.BaseStream` is a `MemoryStream` now instead of a `FileStream` so `Flush` is useless. 
[Because any data written to a MemoryStream object is written into RAM, this method is redundant](https://msdn.microsoft.com/en-us/library/system.io.memorystream.flush(v=vs.110).aspx)
I added `os.Flush` after the `gen` call

But `os` is a BinaryStream so maybe we need to leave that or add `AutoFlush = true` ( see [AutoFlush](https://msdn.microsoft.com/en-us/library/system.io.streamwriter.autoflush%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) ). I'll readd this `os.Flush`, less diff. Good catch :smile:

File: src/absil/ilwrite.fs (hunk 8 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: Sure, i'll try_catch all `FileSystemUtilites.setExecutablePermission` invocations

File: src/absil/ilwrite.fs (hunk 5 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: nice trick!

File: src/absil/ilwrite.fs (hunk 9 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: i'll defend the use of Lazy in a separate comment because i think is usefull and is a strong contract.
But i can inline (as inline refactoring) this function so less confusion ( + a comment where used about the clean used files behaviour )

File: src/absil/ilwrite.fsi (hunk 11 above)
Comment (enricosada) [inline src/absil/ilwrite.fsi]: I'd like to encode contract of  the behaviour in the type system.

for me `(unit -> Stream)` is a problem in the contract api for the Stream:
- can be consumed multiple times, more difficult to call the api (need to dispose multiple streams)
- each call can return a different Stream
- i dont know if it'is already called ( i can track, but it'is easier to already know with `Lazy.IsValueCreated` ).

`Lazy` here is about "do that deferred, but exactly once", i can encode this requirement with a type and a DU, but lazy as type already exists.

About the other improvement, the type

```fsharp
| File of string
| Stream of (unit -> Choice<Stream, Diagnostic>)
```

seem better because i dont expected the string filename to return an error.
An api where the type system enforce that an error is not possible is better. :+1:  I'll use your version

Another way of postpone the problem is to check that at higher level and pass *internally* a non lazy 

```
type EmitTo = File of string | Stream of Stream
```
to all functions below like `writeBinaryAndReportMappings`

- pro: easier code
- pro: the runtime error opening the stream ( `unit -> Choice<Stream,Diagnostic>` ) stop fsc **before** compiling
- cons: the factory ( `unit -> Choice<Stream,Diagnostic>` is called always, but not deferred. The stream is open the whole compilation (not always a bad thing)

@dsyme maybe i can open a discussion in FCS about the api? This is an internal implementation detail, but the problem exists if a public api is exposed, so maybe is better to use same solution.

File: src/absil/ilwrite.fs (hunk 13 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: ok , the `gen` function is like that after several refactoring attempts.
I'd have liked to move this function outside , but the diff was too big, so I left it like that.

I always use inner function because mvoe value binding later, so i can easy see values captured inside big functions like this one.

I will do as you said, less diff.

**Comment by enricosada** [comment]
@dsyme really thx for feedback, i'll fix issues.

I wrote this internal refactoring as less intrusive as possibile, with minimal diff in mind.
I sent the pr there because i'd like a smooth integration with visualfsharp, instead of the difficult merge of multiple commit later when the whole in-memory compilation is ready in fcs.
I really like to minimize merge efforts, it'is wasted time.

I think after this one is merged in visualfsharp, fcs can pull and we can deliver the in-memory compilation without waiting visualfsharp if needed (i'd like to help too).

**Comment by KevinRansom** [comment]
Don, Enrico, mark the pr's to align with [fcs-align] I will prioritise merging them. submit the PRs to master branch.  Please don't break any public API's since the work will go into VS 2015 : update 2.

File: src/absil/ilwrite.fs (hunk 2 above)
Comment (dsyme) [inline src/absil/ilwrite.fs]: Please just keep Choice1Of2/Choice2Of2 and remove the active pattern.  That's in keeping with enough of the style of the rest of the compiler and this PR isn't the right place to change the style. THanks

File: src/absil/ilwrite.fs (hunk 3 above)
Comment (dsyme) [inline src/absil/ilwrite.fs]: Please make this requested change, thanks

Comment (enricosada) [inline src/absil/ilwrite.fs]: @dsyme , sry, my bad, when there is no comment, is ok for me to change it. I'll always write reply next time

File: src/absil/ilwrite.fs (hunk 6 above)
Comment (dsyme) [inline src/absil/ilwrite.fs]: How do we know the upper bound was MAX_PATH before?  I don't see anything in this bit of code limiting to that?

File: src/absil/ilwrite.fsi (hunk 11 above)
Comment (dsyme) [inline src/absil/ilwrite.fsi]: @enricosada Using Lazy in this way is not normal in the compiler.  Code that relies on observing whether lazy values have been created or not is just too subtle.  

I'd rather you somehow made the has-this-stream-been-created information more explicit somehow, rather than relying on Lazy.IsValueCreated.  

thanks

File: src/absil/ilwrite.fs (hunk 8 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: done, is below (because need to be done only for EmittedFile)

```fsharp
path |> trySetExecutablePermission
```

File: src/absil/ilwrite.fsi (hunk 11 above)
Comment (enricosada) [inline src/absil/ilwrite.fsi]: @dsyme ok about remove the Lazy, only question is 
```fsharp
EmittedStream of unit -> Choice<Stream,Diagnostic>
``` 
or directly
```
EmittedStream of Stream
```

i prefer the second, because if the user need a file, should use `EmittedFile`, otherwise should pass directly the `Stream`. 
If the user passa a `FileStream`/`MemoryMappedFile` with `EmittedStream` he should manage the dispose/cleanup (not needed for `MemoryStream` )
Code easier to follow, no need for lazy or deferred creation.

Note that fsc after refactoring use internally  `EmittedFile` not `EmittedStream`

**Comment by enricosada** [comment]
@dsyme done, following your review. 
Now both the diff and implementation is better (less nested), thx

It's open only the `EmittedStream` api, we can wait for a feedback from compiler service contributors i think ( i really like `EmittedStream of Stream` )

**Comment by dsyme** [comment]
Great thanks, I'll take another look

File: src/absil/ilwrite.fs [hunk 20]
```diff
@@ -4025,17 +4074,18 @@ let writeBinaryAndReportMappings (outfile, ilg, pdbfile: string option, signer:
...
           let debugDataJustInCase = 40
+          let MAX_PATH = 260 //windows MAX_PATH, the maximum length of a file path
           let debugDataChunk,next = 
               chunk (align 0x4 (match pdbfile with 
-                                | None -> 0x0 
-                                | Some f -> (24 
-                                            + System.Text.Encoding.Unicode.GetByteCount(f) // See bug 748444
+                                | false -> 0x0 
+                                | true -> (24 
+                                            + MAX_PATH
```
Comment (dsyme) [inline src/absil/ilwrite.fs]: I think I still want this to preserve the old behaviour in the case that we have a named file.  I just don't want us to have any chance of a regression here where we do something wrong with long file names.

File: src/absil/ilwrite.fsi (hunk 11 above)
Comment (dsyme) [inline src/absil/ilwrite.fsi]: Yes, I think I prefer the second.  I'd imagine the final exposed FCS API will just accept a Stream anyway, won't it, and the ultimate caller of FCS will own closing it.

**Comment by dsyme** [comment]
A couple of comments added.  Nice work - the diff is now much cleaner!!

File: src/absil/ilwrite.fs (hunk 20 above)
Comment (enricosada) [inline src/absil/ilwrite.fs]: i'll try to pass the pdb filename as string option, a boolean was easier for me but i think i can do it with string option too.

File: src/absil/ilwrite.fsi (hunk 11 above)
Comment (enricosada) [inline src/absil/ilwrite.fsi]: let's use `EmittedStream of Stream` so. 
We can refactor later in fcs if needed, it's not too bad (less diff here too)

**Comment by enricosada** [comment]
ok, is try to put bak pdb path and use `Stream of Stream`, thx don

File: src/absil/ilwrite.fs (hunk 20 above)
Comment (dsyme) [inline src/absil/ilwrite.fs]: thanks :)

**Comment by enricosada** [comment]
done, less diff, revert to use pdb path instead of MAX_PATH, and cleanups

I'd like to test this code more, so wait a bit before merge ( review is ok :eyes: )

**Comment by dsyme** [comment]
Fantastic! Will take another glance over the code a bit later

**Comment by enricosada** [comment]
ok done testing:
- to stream of pdb/out/debuginfo
- to file of pdb/out/debuginfo
- `--times` flag
- `--test:DumpDebugInfo`

for me is ready to go. 
I'll add unit tests for EmitTo.Stream after the next refactoring (move arguments to `main1`)

**Comment by dsyme** [comment]
OK, this looks good to go to me

**Comment by dsyme** [comment]
(BTW I think I'd expect any further work on the compile-to-stream feature to go to fsharp/FSharp.Compiler.Service after this has been integrated - otherwise it should have been included here?)

**Comment by enricosada** [comment]
I think remains to move parameters up per in the call stack ( `main1` ). I'll add another commit to this pr now that the to stream implementation is done

Inviato da Outlook




On Mon, Nov 16, 2015 at 1:37 PM -0800, "Don Syme" <notifications@github.com> wrote:
(BTW I think I'd expect any further work on the compile-to-stream feature to go to fsharp/FSharp.Compiler.Service after this has been integrated - otherwise it should have been included here?)

---

File: src/fsharp/fsc.fs [hunk 21]
```diff
@@ -1840,30 +1861,38 @@ let main1(tcGlobals, tcImports: TcImports, frameworkTcImports, generatedCcu, typ
       use unwindBuildPhase = PushThreadBuildPhaseUntilUnwind (BuildPhase.Output)    
       if tcConfig.printSignature   then InterfaceFileWriter.WriteInterfaceFile (tcGlobals,tcConfig, InfoReader(tcGlobals,tcImports.GetImportMap()), typedAssembly);
       ReportTime tcConfig ("Write XML document signatures")
-      if tcConfig.xmlDocOutputFile.IsSome then 
+      match xmlDocOutputOpt with
+      | Some(_) ->
```
Comment (enricosada) [inline src/fsharp/fsc.fs]: converted from `if` to `match` because i need type inference, otherwise i must specify method signature for `xmlDocOutputOpt` param (more invasive diff)

**Comment by enricosada** [comment]
ok, i need to do more testing but that should be ok to review at least /cc @dsyme. I'll do more testing asap

I changed only internal functions inside `fsc.fs`, moving all binding one step up in the call stack (`main4` -> `main3` -> .. `main0` ) so the previous review of the others files is ok.

i moved some string path -> EmitTo where possible (like xmldoc)

I cannot move the `outfile` string path upper in the call stack because i need to change `TastPickle.pickleObjWithDanglingCcus` signature inside `EncodeOptimizationData`, but this pr is already big. i can continue later with another.

btw a note, i think we need to change some members of `TcConfig`, like `TcConfig.outputFile` from `string option` to `EmitTo`, so it's easier to propagate the change from `path` to `path or stream`. Typed refactoring is really easier.

After that, we can use a wrapper of `main1` as api in the compiler, so a clean api but the same internal code of the compiler, that should help compiler service (or internally the vs extensions)

I know it's seem useless now, but i really think this refactoring can help with the coreclr because now the code is easier to change (add portable pdb output as another parameter for example)

Each step `main0` -> `main1` -> `main2` -> ... remove arguments, instead of create new ones or local bindings, so it's easier to understand.

File: src/fsharp/fsc.fs [hunk 22]
```diff
@@ -1887,18 +1918,29 @@ let main2(Args(tcConfig, tcImports, frameworkTcImports: TcImports, tcGlobals, er
                                    yield! bytes
                                | _ -> 
                                    failwith "unreachable: expected a local resource" |]
-            let sigDataFileName = (Filename.chopExtension outfile)+".fsdata"
-            File.WriteAllBytes(sigDataFileName,bytes)
+            //TODO instead of bytes array we can directly write to stream, less memory pressure
+            bytes |> emitAllBytes fsDataP
```
Comment (dsyme) [inline src/fsharp/fsc.fs]: FWIW it looks to me like useSignatureDataFile is always false and this is dead code, as is anything to do with fsData files.

**Comment by enricosada** [comment]
i'll rebase to fix emrge conflict

**Comment by KevinRansom** [comment]
@enricosada Enrico, in your opinion is this ready to merge?

**Comment by enricosada** [comment]
Yes, after rebase.

**Comment by enricosada** [comment]
I'll do that soon, this week

**Comment by dsyme** [comment]
Marked WIP until rebase is done (please remove once it is done)

**Comment by dsyme** [comment]
@enricosada Can you bring this up to date with CI tests passing?  It would be good to get it merged. Thanks

**Comment by enricosada** [comment]
@dsyme i'll wait the merge of `coreclr` branch inside `master`. 
I dont want to make the merge more difficult for @KevinRansom because of pdb work

**Comment by KevinRansom** [comment]
@enricosada   coreclr is now merged into master and pretty stable, would you like to take care of this?

Thanks

Kevin

**Comment by KevinRansom** [comment]
@enricosada,

Hey ... do you want to finish this off or shall I close this PR?

Thanks

Kevin

**Comment by enricosada** [comment]
i'll close now and reopen a new pr after a rebase, too much diff atm
